### PR TITLE
Move Teleport binaries from `/usr/local/bin` to `/usr/bin`

### LINKED ISSUE #42130: RPM/DEB Packages incorrectly installing binaries to /usr/local/bin
**Expected behavior**:

Teleport binaries should be installed to `/usr/bin` as per the [Linux Filesystem Hierarchy Standard (FHS)](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html).

**Current behavior**:
Teleport binaries are currently being installed to `/usr/local/bin`, which is intended for manually installed software, not for package-managed software.

**Bug details**:
- Teleport version: 15.3.7
- Recreation steps:
  1. Install Teleport using the RPM/DEB package.
  2. Check the installation directory of the binaries.
- Debug logs: N/A

**Description**:

Hi.

I noticed the RPM and DEB packages for Gravitational Teleport install binaries to `/usr/local/bin`. According to the [Linux Filesystem Hierarchy Standard (FHS)](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html), package-managed software should go to `/usr/bin` instead.

The FHS specifies that `/usr/local` is for manually installed software, not for packages managed by a package manager. Installing to `/usr/local/bin` can cause conflicts with manually installed software.

**Suggestion**:
Could you update the packages to install binaries in `/usr/bin`? This would align with FHS and prevent potential conflicts.

### PR DESCRIPTION
**Pull request by fheinecke**
We currently install Teleport (and related binaries) under `/usr/local/bin`. Per FHS, "The /usr/local hierarchy is for use by the system administrator when installing software locally". On the other hand, FHS says that `/usr/bin` "is the primary directory of executable commands on the system." Only release artifacts that are intended to be installed to be "manually" installed on a system (i.e. tarballs) should be placed under `/usr/local/bin`, and the rest should be under `/usr/bin`.

Release artifacts added in the future should follow the standard as well.

This fixes https://github.com/gravitational/teleport/issues/42130, which is a proposed goal for the internal tools team for next quarter.

Requires https://github.com/gravitational/teleport.e/pull/4478 to be merged at the same time, and an `e` ref update will be needed as well. This will require extensive testing prior to merging. It is my recommendation that we do not backport this, rather, we include this in v17 forward so that all release assets are properly tested prior to pushing to customers. This is arguably a breaking change.

changelog: Changed the install directory for all Teleport binaries from `/usr/local/bin` to `/usr/bin`.

### CODE REVIEW

**Review by tigrato — APPROVED** [review]

**Comment by fheinecke** [comment]
There's a couple of whitespace/linting changes in this PR caused by my IDE. I'll get them reverted prior to merge.

**Review by r0mant — COMMENTED** [review]
Two comments here:

- This feels like it could be a massive breaking change, we should definitely not backport it. Probably risking a bunch of regressions too. Do we want to do this?
- It's a bit difficult to tell from the massive changeset but does this affect only locations of binaries installed via package managers? I think `/usr/local/bin` is the correct location for binaries installed "out of band", and `/usr/bin` is for package-manager-installed programs. The original ticket also only mentions the rpm and deb packages.

**Comment by fheinecke** [comment]
> This feels like it could be a massive breaking change, we should definitely not backport it. Probably risking a bunch of regressions too. Do we want to do this?

IMO "in a vacuum" this is an objectively good change from a technical perspective, as it moves our product in line with a well-defined and accepted industry standard.

That said I think that it is pretty likely that this will introduce some bug(s) somewhere, and should go through a set of major release tests prior to pushing this to customers. I think that it'd be smart to merge this relatively early into the v17 development cycle, as it'll give everybody internally more time to hit these issues and fix them prior to customers finding them.

> It's a bit difficult to tell from the massive changeset but does this affect only locations of binaries installed via package managers? I think /usr/local/bin is the correct location for binaries installed "out of band", and /usr/bin is for package-manager-installed programs. The original ticket also only mentions the rpm and deb packages.

This changes the following in release artifacts and the related supporting files (i.e. docs, service files, etc.):
1. Install paths for debs, rpms, and macos pkg files for all "products" (oss, ent, fips, plugins, updater, etc.).
2. Binary path in AMIs.
3. Binary path in container images.
4. Install scripts in tarballs, but not the binary paths inside tarballs. Note that the install scripts are optional - when we use tarballs in our release pipelines, we ignore the script and just `cp` the binaries.

Artifacts from (1) is directly justified by the FHS and the customer's ticket. Changes from (2) and (3) are correct because we distribute the complete OS (well _almost_ complete OS in the case of container images). The change in (4) is debatable. Personally I think that because the binaries are being installed by something that we produce/manage (the install script) then they should go under `/usr/bin` as well. This also has the advantage of keeping our install path consistent, which makes future regressions less likely.

**Comment by jakule** [comment]
my2c:
1. This is a huge breaking change. If you, for example, create a systemd script using teleport systemd command (or you created it by hand), the `/usr/local/bin` path is hardcoded, and after the upgrade, Teleport will stop working. If you don't have any automated tool, you will need to update those paths in all places for all agents "by hand". We also don't know how much automation our users build around Teleport and how many scripts need to be updated on their side. At that point, I really don't think it's worth changing the path, even if we're not fully complying with the Linux guidelines.
2. If we really, really want to change it we should at least create symlink from the old directory to keep some backward compatibility (symlink won't work for all cases tho).

File: web/packages/teleterm/build_resources/linux/after-remove.tpl
[the remark this answers is not in the document]
Comment (fheinecke) [reply web/packages/teleterm/build_resources/linux/after-remove.tpl:1]: > Should we just extend after-install.tpl to remove the tsh symlink from the old location? We could add a TODO comment to get rid of that in v18.0.0.

We could do this, but might need to leave the logic in through v19... do we support upgrading more than one major version at a time (i.e. v14 to v16)? Can't remember.

[the remark this answers is not in the document]
Comment (avatus) [reply web/packages/teleterm/build_resources/linux/after-remove.tpl:1]: No, we recommend upgrading to [the next major version first](https://goteleport.com/docs/upgrading/overview/#upgrading-a-self-hosted-teleport-cluster) during upgrades

> Upgrade the Auth Service to the next major version first. If there are data format changes introduced in the new version, the Auth Service performs the necessary migrations. After the upgrade, start the Auth Service and CONFIRM that it's in a healthy state before continuing.

**Comment by fheinecke** [comment]
> If you, for example, create a systemd script using teleport systemd command (or you created it by hand), the /usr/local/bin path is hardcoded, and after the upgrade, Teleport will stop working. If you don't have any automated tool, you will need to update those paths in all places for all agents "by hand".

100% agree. This is definitely something that will break customer environments if they don't read the changelog (presumably we will have a nice large entry about this near the top, along with a migration guide).

> At that point, I really don't think it's worth changing the path, even if we're not fully complying with the Linux guidelines.

While I personally think that this work is a good step forward, I completely get this and maybe we shouldn't implement this work. One of the reasons why I filed this as a PR (rather than starting a thread in slack or an issue not directly tied to file changes) was to create a discussion with some context as to they specific impact of this change - not to _necessarily_ get this merged immediately.

> If we really, really want to change it we should at least create symlink from the old directory to keep some backward compatibility (symlink won't work for all cases tho).

Unfortunately this also would violate the parameters of the linked ticket and would continue to violate FHS. Unless we do this _only_ as a temporarily measure for a major version or two, I think that we'd be better off not making this change if it means that we'd have a filesystem entry in both places permanently.

**Comment by zmb3** [comment]
https://github.com/gravitational/teleport/issues/42130#issuecomment-2192703788

**Pull request closed without merging**